daemon/daemon/endpoint.go

```diff
@@ -203,35 +209,67 @@ func (d *Daemon) createBPFMAPs(epID uint16) error {
 	return ep.Regenerate(d)
 }
 
-// EndpointJoin sets up the endpoint working directory.
-func (d *Daemon) EndpointJoin(ep endpoint.Endpoint) error {
-	lxcDir := filepath.Join(".", strconv.Itoa(int(ep.ID)))
+type putEndpoint struct {
+	daemon *Daemon
+}
 
-	if err := os.MkdirAll(lxcDir, 0777); err != nil {
-		log.Warningf("Failed to create container temporary directory: %s", err)
-		return fmt.Errorf("failed to create temporary directory: %s", err)
-	}
+func NewPutEndpointHandler(d *Daemon) PutEndpointHandler {
+	return &putEndpoint{daemon: d}
+}
 
-	d.conf.OptsMU.RLock()
-	ep.SetDefaultOpts(d.conf.Opts)
-	d.conf.OptsMU.RUnlock()
+func (h *putEndpoint) Handle(params PutEndpointParams) middleware.Responder {
+	d := h.daemon
 
-	d.InsertEndpoint(&ep)
+	if params.Endpoint == nil {
+		return NewApiErrorString(401, "No endpoint provided")
+	}
 
-	return nil
-}
+	ep, err := endpoint.NewEndpointFromModel(params.Endpoint)
+	if err != nil {
+		return NewApiError(401, err)
+	}
 
-// EndpointLeave cleans the directory used by the endpoint epID and all relevant details
-// with the epID.
-func (d *Daemon) EndpointLeave(epID uint16) error {
 	d.endpointsMU.Lock()
-	defer d.endpointsMU.Unlock()
+	existingEp, ok := d.endpoints[ep.ID]
```

Comment:
We only read from it but you are right that we have a concurrency issue in general here. We realy on create/join which are decoupled and multiple writers can modify them concurrently. This possibility existed before. I started introducing states to ensure that transitions can only happen in a well defined order but we need to improve this.